docs/config.rst

Default configuration
---------------------

.. note:: Relative path (path which doesn't start by a ``/``) can be used. In this
          case, the absolute path is computed from the configuration file location.

.. code-block:: ini

    [GENERAL]
    # User interface language: 'de', 'dk', 'en', 'es', 'fr', 'hu', 'nl' or 'no'
    language = en

    # Path to save pictures (list of quoted paths accepted)
    directory = ~/Pictures/pibooth

    # Start pibooth at Raspberry Pi startup
    autostart = False

    # In debug mode, exceptions are not caught, logs are more verbose, pictures are cleared at startup
    debug = False

    # Path to custom plugin(s) not installed with pip (list of quoted paths accepted)
    plugins =

    # Plugin names to be disabled after startup (list of quoted names accepted)
    plugins_disabled =

    # Enable a virtual keyboard in the settings interface
    vkeyboard = False

    [WINDOW]
    # The (width, height) of the display window or 'fullscreen'
    size = (800, 480)

    # Background RGB color or image path
    background = (0, 0, 0)

    # Text RGB color
    text_color = (255, 255, 255)

    # Blinking background when a capture is taken
    flash = True

    # Animate the last taken picture by displaying captures one by one
    animate = False

    # How long is displayed the capture in seconds before switching to the next one
    animate_delay = 0.2

    # On 'finish' state: how long is displayed the final picture in seconds (0 if never shown)
    finish_image_delay = 0

    # On 'wait' state: how long is displayed the final picture in seconds before being hidden (-1 if never hidden)
    wait_image_delay = -1

    # Show arrows to indicate physical buttons: 'bottom', 'top', 'hidden' or 'touchscreen'
    arrows = bottom

    # Apply horizontal offset to arrows position
    arrows_x_offset = 0

    # How long is the preview in seconds
    preview_delay = 3

    # Show a countdown timer during the preview
    preview_countdown = True

    # Stop the preview before taking the capture
    preview_stop_on_capture = False

    [PICTURE]
    # Orientation of the final picture: 'auto', 'portrait' or 'landscape'
    orientation = auto

    # Possible choice(s) of captures numbers (numbers between 1 to 4)
    captures = (4, 1)

    # Effect applied to the captures (list of quoted names accepted)
    captures_effects = none

    # Crop each capture border in order to fit the paper size
    captures_cropping = False

    # Thick (in pixels) between captures and picture borders/texts
    margin_thick = 100

    # Main text displayed
    footer_text1 = Footer 1

    # Secondary text displayed
    footer_text2 = Footer 2

    # RGB colors used for footer texts (list of tuples accepted)
    text_colors = (0, 0, 0)

    # Fonts name or file path used for footer texts (list of quoted names accepted)
    text_fonts = ('Amatic-Bold', 'AmaticSC-Regular')

    # Alignments used for footer texts: 'left', 'center' or 'right' (list of quoted names accepted)
    text_alignments = center

    # Overlay path (PNG file) with same aspect ratio than final picture (list of quoted paths accepted)
    overlays =

    # Background RGB color or image path (list of tuples or quoted paths accepted)
    backgrounds = (255, 255, 255)

    [CAMERA]
    # Adjust ISO for lighting issues, can be different for preview and capture (list of integer accepted)
    iso = 100

    # Flip horizontally the capture
    flip = False

    # Rotation of the camera: 0, 90, 180 or 270
    rotation = 0

    # Resolution for camera captures (preview will have same aspect ratio)
    resolution = (1934, 2464)

    # Delete captures from camera internal memory (when applicable)
    delete_internal_memory = False

    [PRINTER]
    # Name of the printer defined in CUPS (or use the 'default' one)
    printer_name = default

    # How long is the print view in seconds (0 to skip it)
    printer_delay = 10

    # Maximum number of printed pages before warning on paper/ink levels (-1 = infinite)
    max_pages = -1

    # Maximum number of duplicate pages sent to the printer (avoid paper waste)
    max_duplicates = 3

    # Print 1, 2, 3 or 4 picture copies per page
    pictures_per_page = 1

    [CONTROLS]
    # How long to press a single hardware button in seconds
    debounce_delay = 0.3

    # How long to press multiple hardware buttons in seconds
    multi_press_delay = 0.5

    # Physical GPIO IN pin to take a picture
    picture_btn_pin = 11

    # Physical GPIO OUT pin to light a LED when picture button is pressed
    picture_led_pin = 7

    # Physical GPIO IN pin to print a picture
    print_btn_pin = 13

    # Physical GPIO OUT pin to light a LED when print button is pressed
    print_led_pin = 15


pibooth/config.py
# -*- coding: utf-8 -*-

"""Pibooth configuration.
"""

import ast
import os
import os.path as osp
import subprocess
from collections import OrderedDict as odict
from pibooth.utils import LOGGER

try:
    from configparser import ConfigParser
except ImportError:
    # Python 2.x fallback
    from ConfigParser import ConfigParser


def get_supported_languages():
    """Return the list of supported language.
    """
    path = osp.join(osp.dirname(osp.abspath(__file__)), 'pictures')
    return [name for name in os.listdir(path) if osp.isdir(osp.join(path, name))
            and not name.startswith('.') and not name.startswith('__')]


def values_list_repr(values):
    """Concatenate a list of values to a readable string.
    """
    return "'{}' or '{}'".format("', '".join([str(i) for i in values[:-1]]), values[-1])


DEFAULT = odict((
    ("GENERAL",
        odict((
            ("language", ("en", "User interface language ({})".format(values_list_repr(get_supported_languages())))),
            ("directory", ("~/Pictures/pibooth", "Path to save pictures")),
            ("clear_on_startup", (True, "Cleanup the 'directory' before start")),
            ("autostart", (False, "Start pibooth at Raspberry Pi startup")),
            ("failsafe", (True, "Show fail message and go back to wait state in case of exception")),
        ))
     ),
    ("WINDOW",
        odict((
            ("size", ((800, 480), "The (width, height) of the display window or 'fullscreen'")),
            ("flash", (True, "Blinking background when picture is taken")),
            ("arrows", (True, "Show arrows to indicate physical buttons")),
            ("preview_delay", (3, "How long is the preview in seconds")),
            ("preview_countdown", (True, "Show a countdown timer during the preview")),
            ("preview_stop_on_capture", (False, "Stop the preview before taking the picture")),
        ))
     ),
    ("PICTURE",
        odict((
            ("captures", ((4, 1), "Possible choice(s) of captures numbers (numbers between 1 to 4 max)")),
            ("orientation", ("auto", "Orientation of the final image ('auto', 'portrait' or 'landscape')")),
            ("effect", ("none", "Effect applied to the captures (a list of quoted names can be given)")),
            ("footer_text1", ("Footer 1", "Main text displayed")),
            ("footer_text2", ("Footer 2", "Secondary text displayed")),
            ("text_color", ((0, 0, 0), "Footer text RGB color")),
            ("bg_color", ((255, 255, 255), "Background RGB color or path to a background image")),
            ("fonts", ("none", "Fonts url to be used on the image")),
        ))
     ),
    ("CAMERA",
        odict((
            ("iso", (100, "Adjust for lighting issues, normal is 100 or 200 and dark is 800 max")),
            ("flip", (False, "Flip horizontally the captured picture")),
            ("rotation", (0, "Rotation of the camera (0, 90, 180 or 270)")),
            ("resolution", ((1934, 2464), "Resolution for camera captures (preview will have same aspect ratio)")),
        ))
     ),
    ("PRINTER",
        odict((
            ("printer_name", ("default", "Name of the printer defined in CUPS (or use the 'default' one)")),
            ("printer_delay", (10, "How long is the print view in seconds (0 to skip it)")),
            ("max_duplicates", (3, "Maximum number of duplicate pages sent to the printer (avoid paper wast)")),
            ("nbr_copies", (1, "Prints 1, 2, 3 or 4 picture copies per page")),
        ))
     ),
    ("CONTROLS",
        odict((
            ("debounce_delay", (0.3, "How long to debounce the hardware buttons in seconds")),
            ("picture_btn_pin", (11, "Physical GPIO IN pin to take a picture")),
            ("picture_led_pin", (7, "Physical GPIO OUT pin to light a LED when picture button is pressed")),
            ("print_btn_pin", (13, "Physical GPIO IN pin to print a picture")),
            ("print_led_pin", (15, "Physical GPIO OUT pin to light a LED when print button is pressed")),
            ("startup_led_pin", (29, "Physical GPIO OUT pin to light a LED at pibooth startup")),
            ("preview_led_pin", (31, "Physical GPIO OUT pin to light a LED during preview")),
        ))
     ),
))


def generate_default_config(filename):
    """Genrate the default configuration.
    """
    with open(filename, 'w') as fp:
        for section, options in DEFAULT.items():
            fp.write("[{}]\n".format(section))
            for name, value in options.items():
                fp.write("# {}\n{} = {}\n\n".format(value[1], name, value[0]))


class PiConfigParser(ConfigParser):

    """Enhenced configuration file parser.
    """

    language = 'en'
    editors = ['leafpad', 'vi', 'emacs']

    def __init__(self, filename, clear=False):
        ConfigParser.__init__(self)
        self.filename = osp.abspath(osp.expanduser(filename))

        if not osp.isfile(self.filename) or clear:
            LOGGER.info("Generate the configuration file in '%s'", self.filename)
            dirname = osp.dirname(self.filename)
            if not osp.isdir(dirname):
                os.makedirs(dirname)
            generate_default_config(self.filename)

        self.reload()

    def reload(self):
        """Reload current configuration file.
        """
        self.read(self.filename)

        # Handle the language configuration, save it as a class attribute for easy access
        language = self.get('GENERAL', 'language')
        if language not in get_supported_languages():
            LOGGER.warning("Unsupported language '%s', fallback to English", language)
        else:
            PiConfigParser.language = language

        # Handle autostart of the application
        self.enable_autostart(self.getboolean('GENERAL', 'autostart'))

    def enable_autostart(self, enable=True):
        """Auto-start pibooth at the Raspberry Pi startup.
        """
        filename = osp.expanduser('~/.config/autostart/pibooth.desktop')
        dirname = osp.dirname(filename)
        if enable and not osp.isfile(filename):

            if not osp.isdir(dirname):
                os.makedirs(dirname)

            LOGGER.info("Generate the auto-startup file in '%s'", dirname)
            with open(filename, 'w') as fp:
                fp.write("[Desktop Entry]\n")
                fp.write("Name=pibooth\n")
                fp.write("Exec=pibooth\n")
                fp.write("Type=application\n")

        elif not enable and osp.isfile(filename):
            LOGGER.info("Remove the auto-startup file in '%s'", dirname)
            os.remove(filename)

    def open_editor(self):
        """Open a text editor to edit the configuration file.
        """
        for editor in self.editors:
            try:
                process = subprocess.Popen([editor, self.filename])
                process.communicate()
                self.reload()
                return
            except OSError as e:
                if e.errno != os.errno.ENOENT:
                    # Something else went wrong while trying to run the editor
                    raise
        LOGGER.critical("Can't find installed text editor among %s", self.editors)

    def get(self, section, option, **kwargs):
        """Override the default function of ConfigParser to add a
        default value if section or option is not found.
        """
        if self.has_section(section) and self.has_option(section, option):
            return ConfigParser.get(self, section, option, **kwargs)
        return str(DEFAULT[section][option][0])

    def gettyped(self, section, option):
        """Get a value from config and try to convert it in a native Python
        type (using the :py:mod:`ast` module).
        """
        value = self.get(section, option)
        try:
            return ast.literal_eval(value)
        except (ValueError, SyntaxError):
            return value

    def getpath(self, section, option):
        """Get a path from config, evaluate the absolute path from configuration
        file path.
        """
        path = self.get(section, option)
        path = osp.expanduser(path)
        if not osp.isabs(path):
            path = osp.join(osp.relpath(osp.dirname(self.filename), '.'), path)
        return path
